### ISSUE TITLE
Add `--pull` when rebuilding without cache to check for base image updates

### ISSUE THREAD

**Issue by rzhao271** [microsoft/vscode-remote-release#5386]
Testing #5373

Using the Node.js config:
On the step that says to run with `--no-cache` (step 4), I get the following output:

```
raymondzhao@Raymonds-MacBook-Pro sandbox-2-nodejs % devcontainer-insiders build --no-cache
[37 ms] Start: Run: docker build -f /Users/raymondzhao/work/sandbox-2-nodejs/.devcontainer/Dockerfile -t vsc-sandbox-2-nodejs-99519735ef2f89be0ef95a34a10d92b6 --no-cache --build-arg VARIANT=16 /Users/raymondzhao/work/sandbox-2-nodejs/.devcontainer
[+] Building 0.2s (5/5) FINISHED                                                
 => [internal] load build definition from Dockerfile                       0.0s
 => => transferring dockerfile: 37B                                        0.0s
 => [internal] load .dockerignore                                          0.0s
 => => transferring context: 2B                                            0.0s
 => [internal] load metadata for mcr.microsoft.com/vscode/devcontainers/j  0.1s
 => CACHED [1/1] FROM mcr.microsoft.com/vscode/devcontainers/javascript-n  0.0s
 => exporting to image                                                     0.0s
 => => exporting layers                                                    0.0s
 => => writing image sha256:a1eff2f1096ac91ea04c1653a8f853e6f9611e20d190e  0.0s
 => => naming to docker.io/library/vsc-sandbox-2-nodejs-99519735ef2f89be0  0.0s

Use 'docker scan' to run Snyk tests against images to find vulnerabilities and learn how to fix them
```

Edit: Using the Node.js and MongoDB config I get the following output:

```
raymondzhao@Raymonds-MacBook-Pro sandbox-2-nodejs-mongodb % devcontainer-insiders build --no-cache
[460 ms] Start: Run: docker-compose --project-name sandbox-2-nodejs-mongodb_devcontainer -f /Users/raymondzhao/work/sandbox-2-nodejs-mongodb/.devcontainer/docker-compose.yml build --no-cache app
Building app
[+] Building 21.1s (7/7) FINISHED                                               
 => [internal] load build definition from Dockerfile                       0.0s
 => => transferring dockerfile: 37B                                        0.0s
 => [internal] load .dockerignore                                          0.0s
 => => transferring context: 2B                                            0.0s
 => [internal] load metadata for mcr.microsoft.com/vscode/devcontainers/j  0.0s
 => CACHED [1/3] FROM mcr.microsoft.com/vscode/devcontainers/javascript-n  0.0s
 => [2/3] RUN curl -sSL "https://www.mongodb.org/static/pgp/server-4.2.a  19.7s
 => [3/3] RUN if [ "1000" != "1000" ] || [ "1000" != "1000" ]; then group  0.3s
 => exporting to image                                                     1.0s 
 => => exporting layers                                                    1.0s 
 => => writing image sha256:124ad56257948560d4a2b36cd240bc7423915d27aa85e  0.0s 
 => => naming to docker.io/library/sandbox-2-nodejs-mongodb_devcontainer_  0.0s 
                                                                                
Use 'docker scan' to run Snyk tests against images to find vulnerabilities and learn how to fix them
```

Therefore, I wonder if it is only the first option, or just Node.js, that is being cached, still.

**Comment by deepak1556** [comment]
Same on linux

```
github/cli-test$ devcontainer-insiders build --no-cache
[24 ms] Start: Run: docker build -f /home/demohan/github/cli-test/.devcontainer/Dockerfile -t vsc-cli-test-f933481297450fd42929702f58e08507 --no-cache --build-arg VARIANT=16 /home/demohan/github/cli-test/.devcontainer
Sending build context to Docker daemon  4.096kB
Step 1/2 : ARG VARIANT="16-buster"
Step 2/2 : FROM mcr.microsoft.com/vscode/devcontainers/javascript-node:0-${VARIANT}
 ---> b01bcb1f2b81
Successfully built b01bcb1f2b81
Successfully tagged vsc-cli-test-f933481297450fd42929702f58e08507:latest
```

**Comment by stuartleeks** [comment]
The `devcontainer` CLI re-uses the extension code for rebuilding without the cache. This code is setting the `--no-cache` option on `docker build` (or `docker-compose build`). With this option, the local cache isn't used for steps in the `Dockerfile`, but if the base image exists locally with a matching tag then it is re-used. 

Looking at the output from @rzhao271, I believe that it is just the base image that is being re-used. We could also pass the `--pull` option which would check for updates to the base image.

**Comment by stuartleeks** [comment]
Update, I've just checked a `docker build --no-cache --pull` and it looks like it still outputs `CACHED` if the base image hasn't been updated and it re-uses it!

```bash
$  docker 'build' --no-cache --pull '-f' '/home/stuart/source/devcontainer-build-run/github-tests/docker-from-docker-non-root/.devcontainer/Dockerfile' '-t' 'vsc-docker-f
rom-docker-non-root-2a0813bdf9b9a9bc2b00f9cd77225f08' '/home/stuart/source/devcontainer-build-run/github-tests/docker-from-docker-non-root/.devcontainer'
[+] Building 81.0s (15/15) FINISHED
 => [internal] load build definition from Dockerfile                                                                                                                       0.0s
 => => transferring dockerfile: 38B                                                                                                                                        0.0s
 => [internal] load .dockerignore                                                                                                                                          0.0s
 => => transferring context: 2B                                                                                                                                            0.0s
 => [internal] load metadata for mcr.microsoft.com/vscode/devcontainers/base:debian-10                                                                                     0.0s
 => CACHED [ 1/10] FROM mcr.microsoft.com/vscode/devcontainers/base:debian-10@sha256:c448230a787689051a6ed68e9673b506f917c5dcccd9a64886c67dec41de0c79                      0.0s
 => [internal] load build context                                                                                                                                          0.0s
 => => transferring context: 110B                                                                                                                                          0.0s
 => [ 2/10] RUN mkdir -p ~/.local/bin                                                                                                                                      0.4s
 => [ 3/10] RUN sudo apt-get update     && sudo apt-get -y install --no-install-recommends apt-utils dialog nano bash-completion sudo bsdmainutils     && sudo apt-get -  20.7s
 => [ 4/10] RUN echo "export HISTFILE=/home/vscode/commandhistory/.bash_history" >> "/home/vscode/.bashrc"     && echo "export PROMPT_COMMAND='history -a'" >> "/home/vsc  0.6s
 => [ 5/10] COPY scripts/golang.sh /tmp/                                                                                                                                   0.0s
 => [ 6/10] RUN /tmp/golang.sh                                                                                                                                            21.0s
 => [ 7/10] RUN     go get github.com/go-delve/delve/cmd/dlv@v1.6.0     && go get golang.org/x/tools/gopls@v0.6.6     && go get github.com/acroca/go-symbols@v0.1.1 && g  24.6s
 => [ 8/10] COPY scripts/docker-client.sh /tmp/                                                                                                                            0.0s
 => [ 9/10] RUN /tmp/docker-client.sh                                                                                                                                     10.7s
 => [10/10] RUN sudo groupadd docker && sudo usermod -aG docker vscode && newgrp docker                                                                                    0.6s
 => exporting to image                                                                                                                                                     2.2s
 => => exporting layers                                                                                                                                                    2.2s
 => => writing image sha256:a06753bd43f74208d6353ae94d86f5a380f8e72a0f049a219cfda63b88071882                                                                               0.0s
 => => naming to docker.io/library/vsc-docker-from-docker-non-root-2a0813bdf9b9a9bc2b00f9cd77225f08                                                                        0.0s
```


Aside from the output, do we want to also have a "Rebuild without cache" check for updates to the base image?

**Comment by stuartleeks** [comment]
Discussed in standup - I will add `--pull` when building with `--no-cache`

**Comment by stuartleeks** [comment]
# Verification steps:

## Setup
- contact @stuartleeks to be added as a contributor to [this repo](https://github.com/stuartleeks/test-pull-on-rebuild) to be able to trigger base image builds. Alternatively, clone and push to a new repo, and fix up the base image name in the dev container definitions.
- Clone https://github.com/stuartleeks/test-pull-on-rebuild

## Steps

The steps below should be verified
 - [ ] with buildkit enabled
 - [ ] without buildkit enabled

To control whether buildkit is enabled, set the following `{ "features": { "buildkit": true } }` property to `true`/`false`. With Docker Desktop this can be found in under Settings/Docker Engine as shown below. (On Linux, this is in `/etc/docker/daemon.json`)

![image](https://user-images.githubusercontent.com/1824461/127327832-8f2eeb24-1913-43f0-a224-9fa92a543a3a.png)

### Dockerfile dev container

1. Open the `dockerfile-test` folder from the cloned repo in VS Code as a dev container
2. In the integrated terminal, run `info` and note the values

```
root@1f7869bea455:/workspace# info
Base date:  Wed Jul 28 12:37:14 UTC 2021
Image date: Wed Jul 28 12:41:06 UTC 2021
root@1f7869bea455:/workspace# 
```

3.  Run the GitHub action [here](https://github.com/stuartleeks/test-pull-on-rebuild/actions/workflows/build-publish-base-image.yml) to build a new version of the base image and wait for it to complete
4. From the Command Palette, run "Remote-Containers: Rebuild Without Cache"
5. In the integrated terminal, run `info` again and verify that the "Base date" has changed

### docker-compose dev container

1. Open the `docker-compose-test` folder from the cloned repo in VS Code as a dev container
2. In the integrated terminal, run `info` and note the values

```
root@1f7869bea455:/workspace# info
Base date:  Wed Jul 28 12:37:14 UTC 2021
Image date: Wed Jul 28 12:41:06 UTC 2021
root@1f7869bea455:/workspace# 
```

3.  Run the GitHub action [here](https://github.com/stuartleeks/test-pull-on-rebuild/actions/workflows/build-publish-base-image.yml) to build a new version of the base image and wait for it to complete
4. From the Command Palette, run "Remote-Containers: Rebuild Without Cache"
5. In the integrated terminal, run `info` again and verify that the "Base date" has changed

**Comment by stuartleeks** [comment]
This is released in `0.190.0`

**Comment by alexr00** [comment]
Thanks for the great verification steps!
